blocks/api/factory.js

```diff
@@ -20,21 +28,28 @@ export function createBlock( name, attributes = {} ) {
 	// Get the type definition associated with a registered block.
 	const blockType = getBlockType( name );
 
-	// Do we need this? What purpose does it have?
-	let defaultAttributes;
-	if ( blockType ) {
-		defaultAttributes = blockType.defaultAttributes;
-	}
+	// Ensure attributes contains only values defined by block type, and merge
+	// default values for missing attributes.
+	attributes = reduce( blockType.attributes, ( result, source, key ) => {
+		const value = attributes[ key ];
+		if ( undefined !== value ) {
+			result[ key ] = value;
+		} else {
+			source = getNormalizedAttributeSource( source );
```

Comment:
Have you considered calling this function (`getNormalizedAttributeSource`) when registering the block to avoid having to do it everytime we need the block attributes?